data/abi/abi_type.go

```diff
@@ -0,0 +1,491 @@
+            | <type> [<N>]
+            | <type> []
+            | string
+            | (T1, ..., Tn)
+*/
+
+// BaseType is an type-alias for uint32. A BaseType value indicates the type of an ABI value.
+type BaseType uint32
+
+const (
+	// Uint is the index (0) for `Uint` type in ABI encoding.
+	Uint BaseType = iota
+	// Byte is the index (1) for `Byte` type in ABI encoding.
+	Byte
+	// Ufixed is the index (2) for `UFixed` type in ABI encoding.
+	Ufixed
+	// Bool is the index (3) for `Bool` type in ABI encoding.
+	Bool
+	// ArrayStatic is the index (4) for static length array (<type>[length]) type in ABI encoding.
+	ArrayStatic
+	// Address is the index (5) for `Address` type in ABI encoding (an type alias of Byte[32]).
+	Address
+	// ArrayDynamic is the index (6) for dynamic length array (<type>[]) type in ABI encoding.
+	ArrayDynamic
+	// String is the index (7) for `String` type in ABI encoding (an type alias of Byte[]).
+	String
+	// Tuple is the index (8) for tuple `(<type 0>, ..., <type k>)` in ABI encoding.
+	Tuple
+)
+
+// Type is the struct that stores information about an ABI value's type.
+type Type struct {
+	abiTypeID  BaseType
+	childTypes []Type
+
+	// only can be applied to `uint` bitSize <N> or `ufixed` bitSize <N>
+	bitSize uint16
+	// only can be applied to `ufixed` precision <M>
+	precision uint16
+
+	// length for static array / tuple
+	/*
+		by ABI spec, len over binary array returns number of bytes
+		the type is uint16, which allows for only lenth in [0, 2^16 - 1]
+		representation of static length can only be constrained in uint16 type
+	*/
+	// NOTE may want to change back to uint32/uint64
+	staticLength uint16
+}
+
+// String serialize an ABI Type to a string in ABI encoding.
+func (t Type) String() string {
+	switch t.abiTypeID {
+	case Uint:
+		return "uint" + strconv.Itoa(int(t.bitSize))
+	case Byte:
+		return "byte"
+	case Ufixed:
+		return "ufixed" + strconv.Itoa(int(t.bitSize)) + "x" + strconv.Itoa(int(t.precision))
+	case Bool:
+		return "bool"
+	case ArrayStatic:
+		return t.childTypes[0].String() + "[" + strconv.Itoa(int(t.staticLength)) + "]"
+	case Address:
+		return "address"
+	case ArrayDynamic:
+		return t.childTypes[0].String() + "[]"
+	case String:
+		return "string"
+	case Tuple:
+		typeStrings := make([]string, len(t.childTypes))
+		for i := 0; i < len(t.childTypes); i++ {
+			typeStrings[i] = t.childTypes[i].String()
+		}
+		return "(" + strings.Join(typeStrings, ",") + ")"
+	default:
+		panic("Type Serialization Error, fail to infer from abiTypeID (bruh you shouldn't be here)")
+	}
+}
+
+var staticArrayRegexp *regexp.Regexp = nil
+var ufixedRegexp *regexp.Regexp = nil
+
+func init() {
+	var err error
+	// Note that we allow only decimal static array length
+	staticArrayRegexp, err = regexp.Compile(`^([a-z\d\[\](),]+)\[([1-9][\d]*)]$`)
+	if err != nil {
+		panic(err.Error())
+	}
+	ufixedRegexp, err = regexp.Compile(`^ufixed([1-9][\d]*)x([1-9][\d]*)$`)
+	if err != nil {
+		panic(err.Error())
```

Comment:
Simplify with https://pkg.go.dev/regexp#MustCompile ?  I suppose that could even make it doable outside init() at global level.

Reply:
I removed the init function, put the regexp compile to global level.